README.md
json2xls:Generate Excel by JSON data
====================================

[![](https://badge.fury.io/py/json2xls.png)](http://badge.fury.io/py/json2xls)
[![](https://pypip.in/d/json2xls/badge.png)](https://pypi.python.org/pypi/json2xls)

       _                 ____       _
      (_)___  ___  _ __ |___ \__  _| |___
      | / __|/ _ \| '_ \  __) \ \/ / / __|
      | \__ \ (_) | | | |/ __/ >  <| \__ \
     _/ |___/\___/|_| |_|_____/_/\_\_|___/
    |__/

generate excel by json string or json file or url which return a json

**docs** <http://json2xls.readthedocs.org/en/latest/>

**install**

    pip install json2xls

or

    python setup.py install


## command usage:

#### gen xls from json string

    json2xls tests/cmd_str_test.xls '{"a":"a", "b":"b"}'
    json2xls tests/cmd_str_test1.xls '[{"a":"a", "b":"b"},{"a":1, "b":2}]'

#### gen xls from file: whole file is a complete json data

    json2xls tests/cmd_list_test.xls "`cat tests/list_data.json`"

#### gen xls from file: each line is a json data

    json2xls tests/cmd_line_test.xls tests/line_data.json

#### gen xls from a url which respond a json

    json2xls tests/cmd_get_test.xls http://httpbin.org/get
    json2xls tests/cmd_post_test.xls http://httpbin.org/post -m post -d '"hello json2xls"' -h "{'X-Token': 'bolobolomi'}"

## coding usage:

#### gen xls from json string

    #!/usr/bin/env python
    #-*- coding:utf-8 -*-

    from json2xls.json2xls import Json2Xls

    json_data = u'''[
        {"姓名": "John", "年龄": 30, "性别": "男"},
        {"姓名": "Alice", "年龄": 18, "性别": "女"}
    ]'''
    obj = Json2Xls('tests/json_strlist_test.xls', json_data)
    obj.make()


#### gen xls from a url which respond a json by GET

    params = {
        'location': u'上海',
        'output': 'json',
        'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
    }
    Json2Xls('tests/url_get_test.xls',
             "http://httpbin.org/get",
             params=params).make()


#### gen xls from a url which respond a json by POST

    post_data = {
        'location': u'上海',
        'output': 'json',
        'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
    }
    Json2Xls('tests/url_post_test1.xls',
             "http://httpbin.org/post",
             method='post',
             post_data=post_data,
             form_encoded=True).make()

#### gen xls from a url which respond a json by POST with file post data

    post_data = 'tests/post_data.json'
    Json2Xls('tests/url_post_test2.xls',
             "http://httpbin.org/post",
             method='post',
             post_data=post_data,
             form_encoded=True).make()


#### gen xls from file: whole file is a complete json data （从文件内容为一个json列表的文件生成excel）

    Json2Xls('tests/json_list_test.xls', json_data='tests/list_data.json').make()

#### gen xls from file: each line is a json data （从文件内容为每行一个的json字符串的文件生成excel）

    obj = Json2Xls('tests/json_line_test.xls', json_data='tests/line_data.json')
    obj.make()

#### gen custom excel by define your title and body callback function

    def title_callback(self, data):
        '''use one of data record to generate excel title'''
        self.sheet.write_merge(0, 0, 0, 3, 'title', self.title_style)
        self.sheet.write_merge(1, 2, 0, 0, 'tag', self.title_style)
        self.sheet.write_merge(1, 2, 1, 1, 'ner', self.title_style)
        self.sheet.write_merge(1, 1, 2, 3, 'comment', self.title_style)
        self.sheet.row(2).write(2, 'x', self.title_style)
        self.sheet.row(2).write(3, 'y', self.title_style)

        self.sheet.write_merge(0, 0, 4, 7, 'body', self.title_style)
        self.sheet.write_merge(1, 2, 4, 4, 'tag', self.title_style)
        self.sheet.write_merge(1, 2, 5, 5, 'ner', self.title_style)
        self.sheet.write_merge(1, 1, 6, 7, 'comment', self.title_style)
        self.sheet.row(2).write(6, 'x', self.title_style)
        self.sheet.row(2).write(7, 'y', self.title_style)

        self.start_row += 3

    def body_callback(self, data):

        key1 = ['title', 'body']
        key2 = ['tag', 'ner', 'comment']

        col = 0
        for ii, i in enumerate(key1):
            for ij, j in enumerate(key2):
                if j != 'comment':
                    value = ', '.join(data[ii][i][j])
                    self.sheet.row(self.start_row).write(col, value)
                    col += 1
                else:
                    for x in data[ii][i][j].values():
                        self.sheet.row(self.start_row).write(col, x)
                        col += 1
        self.start_row += 1

    data = 'tests/callback_data.json'
    j = Json2Xls('tests/callback.xls', data)
    j.make(title_callback=title_callback, body_callback=body_callback)


Default request method is `get`, request argument pass by `params`.
and the `post` method's request argument pass by `data`, you can use `-d` to pass request data in command line, the data should be json or file

Default only support one layer json to generate the excel, the nested json will be flattened. if you want custom it,
you can write the `title_callback` function and `body_callback` function, the pass them in the `make` function.
for the `body_callback`, you just need to care one line data's write way, json2xls default think the data are all the same.

The test demo data file is in tests dir.


HISTORY.rst
.. :changelog:

History
-------

0.0.2 (2014-09-24)
---------------------

* add callback for no dict depth limit
* update title style
* col width auto adjust
* fixed headers bug
* support json of line in file


0.0.1 (2014-09-23)
---------------------

* First release on PyPI.


tests/tests.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

from json2xls import Json2Xls

url_or_json = '''[
    {"name": "John", "age": 30, "sex": "male"},
    {"name": "Alice", "age": 18, "sex": "female"}
]'''
obj = Json2Xls('test.xls', url_or_json)
obj.make()

params = {
    'location': u'上海',
    'output': 'json',
    'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
}
Json2Xls('test2.xls', "http://api.map.baidu.com/telematics/v3/weather", params=params).make()




def title_callback(self, data):
    '''use one data record to generate excel title'''
    self.sheet.write_merge(0, 0, 0, 3, 'title', self.title_style)
    self.sheet.write_merge(1, 2, 0, 0, 'tag', self.title_style)
    self.sheet.write_merge(1, 2, 1, 1, 'ner', self.title_style)
    self.sheet.write_merge(1, 1, 2, 3, 'comment', self.title_style)
    self.sheet.row(2).write(2, 'x', self.title_style)
    self.sheet.row(2).write(3, 'y', self.title_style)

    self.sheet.write_merge(0, 0, 4, 7, 'body', self.title_style)
    self.sheet.write_merge(1, 2, 4, 4, 'tag', self.title_style)
    self.sheet.write_merge(1, 2, 5, 5, 'ner', self.title_style)
    self.sheet.write_merge(1, 1, 6, 7, 'comment', self.title_style)
    self.sheet.row(2).write(6, 'x', self.title_style)
    self.sheet.row(2).write(7, 'y', self.title_style)

    self.start_row += 3


def body_callback(self, data):

    key1 = ['title', 'body']
    key2 = ['tag', 'ner', 'comment']

    col = 0
    for ii, i in enumerate(key1):
        for ij, j in enumerate(key2):
            if j != 'comment':
                value = ', '.join(data[ii][i][j])
                self.sheet.col(col).width = (len(value) + 1) * 256
                self.sheet.row(self.start_row).write(col, value)
                col += 1
            else:
                for x in data[ii][i][j].values():
                    width = self.sheet.col(col).width
                    new_width = (len(x) + 1) * 256
                    self.sheet.col(col).width = width if width > new_width else new_width
                    self.sheet.row(self.start_row).write(col, x)
                    col += 1
    self.start_row += 1


data = '''[
            [
                {
                    "title":
                        {
                            "tag": ["title_tag1", "title_tag2"],
                            "ner": ["title_ner1", "title_ner2"],
                            "comment": { "good": "100", "bad": "20"}
                        }
                },
                {
                    "body":
                        {
                            "tag": ["body_tag1", "body_tag2"],
                            "ner": ["body_ner1", "body_ner2"],
                            "comment": { "good": "85", "bad": "60"}
                        }
                }
            ]
        ]'''

j = Json2Xls('title_callback.xls', data)
j.make(title_callback=title_callback, body_callback=body_callback)


tests/test_json2xls.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_json2xls
----------------------------------

Tests for `json2xls` module.
"""

import unittest

from json2xls import json2xls


class TestJson2xls(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

demo.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

from json2xls.json2xls import Json2Xls

json_data = u'''[
    {"姓名": "John", "年龄": 30, "性别": "男"},
    {"姓名": "Alice", "年龄": 18, "性别": "女"}
]'''
obj = Json2Xls('tests/json_strlist_test.xls', json_data)
obj.make()


params = {
    'location': u'上海',
    'output': 'json',
    'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
}
Json2Xls('tests/url_get_test.xls', "http://httpbin.org/get", params=params).make()


obj = Json2Xls('tests/json_list_test.xls', json_data='tests/list_data.json')
obj.make()


obj = Json2Xls('tests/json_line_test.xls', json_data='tests/line_data.json')
obj.make()


post_data = {
    'location': u'上海',
    'output': 'json',
    'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
}
Json2Xls('tests/url_post_test1.xls', "http://httpbin.org/post", method='post', post_data=post_data, form_encoded=True).make()


post_data = 'tests/post_data.json'
Json2Xls('tests/url_post_test2.xls', "http://httpbin.org/post", method='post', post_data=post_data, form_encoded=True).make()


def title_callback(self, data):
    '''use one of data record to generate excel title'''
    self.sheet.write_merge(0, 0, 0, 3, 'title', self.title_style)
    self.sheet.write_merge(1, 2, 0, 0, 'tag', self.title_style)
    self.sheet.write_merge(1, 2, 1, 1, 'ner', self.title_style)
    self.sheet.write_merge(1, 1, 2, 3, 'comment', self.title_style)
    self.sheet.row(2).write(2, 'x', self.title_style)
    self.sheet.row(2).write(3, 'y', self.title_style)

    self.sheet.write_merge(0, 0, 4, 7, 'body', self.title_style)
    self.sheet.write_merge(1, 2, 4, 4, 'tag', self.title_style)
    self.sheet.write_merge(1, 2, 5, 5, 'ner', self.title_style)
    self.sheet.write_merge(1, 1, 6, 7, 'comment', self.title_style)
    self.sheet.row(2).write(6, 'x', self.title_style)
    self.sheet.row(2).write(7, 'y', self.title_style)

    self.start_row += 3

def body_callback(self, data):

    key1 = ['title', 'body']
    key2 = ['tag', 'ner', 'comment']

    col = 0
    for ii, i in enumerate(key1):
        for ij, j in enumerate(key2):
            if j != 'comment':
                value = ', '.join(data[ii][i][j])
                self.sheet.row(self.start_row).write(col, value)
                col += 1
            else:
                for x in data[ii][i][j].values():
                    self.sheet.row(self.start_row).write(col, x)
                    col += 1
    self.start_row += 1

data = 'tests/callback_data.json'
j = Json2Xls('tests/callback.xls', data)
j.make(title_callback=title_callback, body_callback=body_callback)



json2xls/json2xls.py
# !/usr/bin/env python
# -*- coding:utf-8 -*-
import collections
import json
import os
from collections import OrderedDict
from functools import partial

import click
import requests

import xlwt
from xlwt import Workbook


class Json2Xls(object):
    """Json2Xls API 接口

    :param string xls_filename: 指定需要生成的的excel的文件名

    :param string json_data: 指定json数据来源，
       可以是一个返回json的url，
       也可以是一行json字符串，
       也可以是一个包含每行一个json的文本文件

    :param string method: 当数据来源为url时，请求url的方法，
       默认为get请求

    :param dict params: get请求参数，默认为 :py:class:`None`

    :param dict post_data: post请求参数，默认为 :py:class:`None`

    :param dict headers: 请求url时的HTTP头信息 (字典、json或文件)

    :param bool form_encoded: post请求时是否作为表单请求，默认为 :py:class:`False`

    :param string sheet_name: Excel的sheet名称，默认为 sheet0

    :param string title_style: Excel的表头样式，默认为 :py:class:`None`

    :param function json_dumps: 带ensure_ascii参数的json.dumps()，
                                默认参数值为 :py:class:`False`

    :param function json_loads: 带object_pairs_hook参数的json.loads()，默认为保持原始排序

    :param bool dumps: 生成excel时对表格内容执行json_dumps，默认为 :py:class:`False`
    """

    def __init__(self, xls_filename, json_data,
                 method='get',
                 params=None,
                 post_data=None,
                 headers=None,
                 form_encoded=False,
                 dumps=False,
                 sheet_name='sheet0',
                 title_style=None):
        self.json_dumps = partial(json.dumps, ensure_ascii=False)
        self.json_loads = partial(json.loads, object_pairs_hook=OrderedDict)

        self.sheet_name = sheet_name
        self.xls_filename = xls_filename
        self.json_data = json_data
        self.method = method
        self.params = params
        self.post_data = post_data
        self.headers = headers
        self.form_encoded = form_encoded
        self.dumps = dumps

        self.__check_file_suffix()

        self.book = Workbook(encoding='utf-8', style_compression=2)
        self.sheet = self.book.add_sheet(self.sheet_name)

        self.start_row = 0

        self.title_style = xlwt.easyxf(
            title_style or 'font: name Arial, bold on;'
            'align: vert centre, horiz center;'
            'borders: top 1, bottom 1, left 1, right 1;'
            'pattern: pattern solid, fore_colour lime;')

    def __check_file_suffix(self):
        suffix = self.xls_filename.split('.')[-1]
        if '.' not in self.xls_filename:
            self.xls_filename += '.xls'
        elif suffix != 'xls':
            raise Exception('filename suffix must be .xls')

    def __get_json(self):
        data = None
        try:
            data = self.json_loads(self.json_data)
        except:
            if os.path.isfile(self.json_data):
                with open(self.json_data, 'r') as source:
                    try:
                        data = self.json_loads(
                            source.read().decode('utf-8').replace('\n', ''))
                    except:
                        source.seek(0)
                        data = [self.json_loads(line.decode('utf-8'))
                                for line in source]
            else:
                if os.path.isfile(self.headers):
                    with open(self.headers) as headers_txt:
                        self.headers = self.json_loads(
                            headers_txt.read().decode('utf-8').replace('\n',
                                                                       ''))
                elif isinstance(self.headers, basestring):
                    self.headers = self.json_loads(self.headers)
                try:
                    if self.method.lower() == 'get':
                        resp = requests.get(self.json_data,
                                            params=self.params,
                                            headers=self.headers)
                        data = resp.json()
                    else:
                        if isinstance(self.post_data,
                                      basestring) and os.path.isfile(
                                          self.post_data):
                            with open(self.post_data, 'r') as source:
                                self.post_data = self.json_loads(
                                    source.read().decode('utf-8').replace('\n',
                                                                          ''))
                        if not self.form_encoded:
                            self.post_data = self.json_dumps(self.post_data)
                        resp = requests.post(self.json_data,
                                             data=self.post_data,
                                             headers=self.headers)
                        data = resp.json()
                except Exception as e:
                    print e
        return data

    def __fill_title(self, data):
        '''生成默认title'''
        data = self.flatten(data)
        for index, key in enumerate(data.keys()):
            if self.dumps:
                key = self.json_dumps(key)
            try:
                self.sheet.col(index).width = (len(key) + 1) * 256
            except:
                pass
            self.sheet.row(self.start_row).write(index, key, self.title_style)
        self.start_row += 1

    def __fill_data(self, data):
        '''生成默认sheet'''
        data = self.flatten(data)
        for index, value in enumerate(data.values()):
            if self.dumps:
                value = self.json_dumps(value)
            self.auto_width(self.start_row, index, value)
            self.sheet.row(self.start_row).write(index, value)

        self.start_row += 1

    def auto_width(self, row, col, value):
        '''单元格宽度自动伸缩

        :param int row: 单元格所在行下标

        :param int col: 单元格所在列下标

        :param int value: 单元格中的内容
        '''

        try:
            self.sheet.row(row).height_mismatch = True
            # self.sheet.row(row).height = 0
            width = self.sheet.col(col).width
            new_width = min((len(value) + 1) * 256, 256 * 50)
            self.sheet.col(col).width = width \
                if width > new_width else new_width
        except:
            pass

    def flatten(self, data_dict, parent_key='', sep='.'):
        '''对套嵌的dict进行flatten处理为单层dict

        :param dict data_dict: 需要处理的dict数据。

        :param str parent_key: 上层字典的key，默认为空字符串。

        :param str sep: 套嵌key flatten后的分割符， 默认为“.” 。
        '''

        out = {}

        def _flatten(x, parent_key, sep):
            if isinstance(x, collections.MutableMapping):
                for a in x:
                    _flatten(x[a], parent_key + a + sep, sep)
            elif isinstance(x, collections.MutableSequence):
                i = 0
                for a in x:
                    _flatten(a, parent_key + str(i) + sep, sep)
                    i += 1
            else:
                out[str(parent_key[:-1])] = str(x)
        _flatten(data_dict, parent_key, sep)
        return OrderedDict(out)

    def make(self, title_callback=None, body_callback=None):
        '''生成Excel。

        :param func title_callback: 自定义生成Execl表头的回调函数。
           默认为 :py:class:`None`，即采用默认方法生成

        :param func body_callback: 自定义生成Execl内容的回调函数。
           默认为 :py:class:`None`，即采用默认方法生成
        '''

        data = self.__get_json()
        if not isinstance(data, (dict, list)):
            raise Exception('The %s is not a valid json format' % type(data))
        if isinstance(data, dict):
            data = [data]

        if title_callback is not None:
            title_callback(self, data[0])
        else:
            self.__fill_title(data[0])

        if body_callback is not None:
            for d in data:
                body_callback(self, d)
        else:
            for d in data:
                self.__fill_data(d)

        self.book.save(self.xls_filename)


@click.command()
@click.argument('xls_filename')
@click.argument('json_data')
@click.option('--method', '-m', default='get')
@click.option('--params', '-p', default=None)
@click.option('--post_data', '-d', default=None)
@click.option('--headers', '-h', default=None)
@click.option('--sheet', '-s', default='sheet0')
@click.option('--style', '-S', default=None)
@click.option('--form_encoded', '-f', is_flag=True)
@click.option('--dumps', '-D', is_flag=True)
def make(xls_filename, json_data, method, params, post_data, headers, sheet,
         style, form_encoded, dumps):
    Json2Xls(xls_filename, json_data,
             method=method,
             params=params,
             post_data=post_data,
             headers=headers,
             form_encoded=form_encoded,
             dumps=dumps,
             sheet_name=sheet,
             title_style=style).make()


if __name__ == '__main__':
    make()


json2xls/__init__.py
#!/usr/bin/env python
# coding:utf-8

"""
json2xls
===========

根据json数据生成excel表格，默认支持生成单层json的Excel，多层json会默认被转化为单层，也可以自定义多层样式的生成方法。

json数据来源可以是一个返回json的url，也可以是一行json字符串，也可以是一个包含每行一个json的文本文件

安装
----

:py:mod:`json2xls` 代码托管在 `GitHub`_，并且已经发布到 `PyPI`_，可以直接通过 `pip` 安装::

    $ pip install json2xls

源码安装::

    $ python setup.py install

:py:mod:`json2xls` 以 MIT 协议发布。

.. _GitHub: https://github.com/axiaoxin/json2xls
.. _PyPI: https://pypi.python.org/pypi/json2xls

使用教程
--------

API调用::

    from json2xls import Json2Xls

    # 从json字符串生成excel
    json_data = u'''[
        {"姓名": "John", "年龄": 30, "性别": "男"},
        {"姓名": "Alice", "年龄": 18, "性别": "女"}
    ]'''
    obj = Json2Xls('tests/json_strlist_test.xls', json_data)
    obj.make()


    # 从get请求返回的json生成excel
    params = {
        'location': u'上海',
        'output': 'json',
        'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
    }
    Json2Xls('tests/url_get_test.xls', "http://httpbin.org/get", params=params).make()


    # 从post请求返回的json生成excel
    post_data = {
        'location': u'上海',
        'output': 'json',
        'ak': '5slgyqGDENN7Sy7pw29IUvrZ'
    }
    Json2Xls('tests/url_post_test1.xls', "http://httpbin.org/post", method='post', post_data=post_data, form_encoded=True).make()
    # 如果post_data很复杂很长可以写到一个文件里
    post_data = 'tests/post_data.json'
    Json2Xls('tests/url_post_test2.xls', "http://httpbin.org/post", method='post', post_data=post_data, form_encoded=True).make()


    # 从文件内容为每行一个的json字符串的文件生成excel
    obj = Json2Xls('tests/json_line_test.xls', json_data='tests/line_data.json')
    obj.make()
    # 从文件内容为一个json列表的文件生成excel
    Json2Xls('tests/json_list_test.xls', json_data='tests/list_data.json').make()

    # 自定义生成excel
    def title_callback(self, data):
        '''use one of data record to generate excel title'''
        self.sheet.write_merge(0, 0, 0, 3, 'title', self.title_style)
        self.sheet.write_merge(1, 2, 0, 0, 'tag', self.title_style)
        self.sheet.write_merge(1, 2, 1, 1, 'ner', self.title_style)
        self.sheet.write_merge(1, 1, 2, 3, 'comment', self.title_style)
        self.sheet.row(2).write(2, 'x', self.title_style)
        self.sheet.row(2).write(3, 'y', self.title_style)

        self.sheet.write_merge(0, 0, 4, 7, 'body', self.title_style)
        self.sheet.write_merge(1, 2, 4, 4, 'tag', self.title_style)
        self.sheet.write_merge(1, 2, 5, 5, 'ner', self.title_style)
        self.sheet.write_merge(1, 1, 6, 7, 'comment', self.title_style)
        self.sheet.row(2).write(6, 'x', self.title_style)
        self.sheet.row(2).write(7, 'y', self.title_style)

        self.start_row += 3

    def body_callback(self, data):

        key1 = ['title', 'body']
        key2 = ['tag', 'ner', 'comment']

        col = 0
        for ii, i in enumerate(key1):
            for ij, j in enumerate(key2):
                if j != 'comment':
                    value = ', '.join(data[ii][i][j])
                    self.sheet.row(self.start_row).write(col, value)
                    col += 1
                else:
                    for x in data[ii][i][j].values():
                        self.sheet.row(self.start_row).write(col, x)
                        col += 1
        self.start_row += 1

    data = 'tests/callback_data.json'
    j = Json2Xls('tests/callback.xls', data)
    j.make(title_callback=title_callback, body_callback=body_callback)

命令行::

    # from json string
    json2xls tests/cmd_str_test.xls '{"a":"a", "b":"b"}'
    json2xls tests/cmd_str_test1.xls '[{"a":"a", "b":"b"},{"a":1, "b":2}]'

    # from file: whole file is a complete json data
    json2xls tests/cmd_list_test.xls "`cat tests/list_data.json`"

    # from file: each line is a json data
    json2xls tests/cmd_line_test.xls tests/line_data.json

    # from url
    json2xls tests/cmd_get_test.xls http://httpbin.org/get
    json2xls tests/cmd_post_test.xls http://httpbin.org/post -m post -d '"hello json2xls"' -h "{'X-Token': 'bolobolomi'}"

"""

__author__ = 'axiaoxin'
__email__ = '254606826@qq.com'
__version__ = '0.1.3c'

from json2xls import Json2Xls
